### PR TITLE
Add support for unknown aggregators and post-aggregators

### PR DESCRIPTION
**Pull request by Fryuni**
Since Druid extensions may define aggregators and post-aggregators, it is infeasible for this library to individually support every possible aggregation type.
Even more so, two extensions might define different aggregations with the same name.

This PR adds support for unknown aggregation and post-aggregation types by providing the types `aggregation.Generic` and `postaggregation.Generic`, which are backed by a `map[string]interface{}`.

Extensions may also define any kind of component, not just aggregations and post-aggregations, but this PR is only for those two component kinds. I might open separate PRs to support generic queries, filters and datasources.

### CODE REVIEW

**Comment by Fryuni** [comment]
Ok, I'll change the name.

**Comment by Fryuni** [comment]
> Then if there is a type "JSON" (or, "generic", same reasoning) , we still should keep the default case returning an error "unknown type" instead of making JSON/Generic the default.

"JSON" is not the type, the point is exactly that there can be any unknown types that are valid. For example this is a valid aggregation when using [Tuple Sketches](https://druid.apache.org/docs/latest/development/extensions-core/datasketches-tuple.html):
```json
{
  "type" : "arrayOfDoublesSketch",
  "name" : "foo",
  "fieldName" : "bar",
  "nominalEntries": 1024,
  "metricColumns" : [],
  "numberOfValues" : 0
 }
```

The `type` is not in the list of known types, that is when the new JSON component comes in. It needs to Marshal to and Unmarshal from the JSON above.

**Comment by Fryuni** [comment]
Any extension, even private ones can define those new components. We have, for example, an extension of that extension where we add a variation of one of its post-aggregations.

```json
{
  "type": "arrayOfDoublesSketchSetOpWithOperation", <== This type only exists in our cluster
  "name": <output name>,
  "combineOperation": ...,   <==== Because we add this field
  "operation": <"UNION"|"INTERSECT"|"NOT">,
  "fields": ...,
  "nominalEntries": ...,
  "numberOfValues": ...
}
```

So even if types for the Tuple Sketch extension are added explicitly, it would not work for the JSON above without defaulting to a generic JSON component.

**Comment by jbguerraz** [comment]
True! I've been too fast)

File: builder/json.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package builder
+
+import "encoding/json"
+
+type JSON struct {
+	Typ    ComponentType
+	Fields map[string]interface{}
```
Comment (jbguerraz) [inline builder/json.go:7]: maybe let's name it spec ?

Comment (Fryuni) [reply builder/json.go:7]: Some options:
- `JSON`/`JSONComponent`
- back to `Generic`/`GenericComponent`
- `Spec`/`SpecComponent` (I don't quite like it)
- `Custom`/`CustomComponent`
- `NonStandardComponent`
- `Any`/`AnyComponent`

I like the `Custom` and the `AnyComponent`, WDYT?

Comment (jbguerraz) [reply builder/json.go:7]: I'd name it spec because that's the Druid semantic https://github.com/search?q=repo%3Aapache%2Fdruid%20spec&type=code

Comment (Fryuni) [reply builder/json.go]: That makes sense. Done!

**Comment by Fryuni** [comment]
@jbguerraz ping

Should I do anything else here?

**Review by jbguerraz — APPROVED** [review]
LGTM! thx @Fryuni

**Pull request merged**